### DISCUSSION TITLE
[A-L] (23 luglio 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#112] [Esami A-L]
**Esercizio 3:** dare lo pseudocodice di un algoritmo che accetta in input un grafo $G$ diretto e aciclico con vertici colorati rosso, blu e verde e una coppia di vertici $u$ e $v$ . L'algoritmo determina se esiste un cammino diretto da $u$ a $v$ con vertici di tutti e tre i colori presenti nel cammino. (Quindi, se esiste un camino con al minimo un vertice rosso, al minimo uno blu, e al minimo uno verde, l'algoritmo restituisce "SI", altrimenti, se ogni cammino da $u$ a $v$ ha al massimo due colori distinti, l'algoritmo restituisce "NO"). 
L'algoritmo dovrebbe avere complessità $O(n+m)$ .

**Comment by Elia-Belli** [comment]
## DFS (Soluzione Errata ❌ )
Possiamo eseguire una DFS, memorizzando durante la visita in profondità i colori incontrati in delle variabili:
- se raggiungiamo $v$ e tutti i colori sono stati incontrati allora il cammino esiste
- altrimenti giunti ad una foglia dell'arborescenza decrementiamo la variabile del suo colore prima di tornare dalla chiamata ricorsiva, per non falsificare i colori raggiunti nella prossima visita in profondità, e proseguiamo allo stesso modo.

Essendo una DFS la complessità è $O(n+m)$

```python
def colored_path(G,C,u,v):
    # rgb: numero di nodi di ciascun colore nel cammino da u al nodo attualmente visitato
    rgb = {'red':0, 'green':0, 'blue':0}
    rgb[C[v]] += 1

    P = [-1]*n
    P[u] = u

    return DFS_color(G,C,P,rgb,u,v)


def DFS_color(G,C,P,rgb,u,v):
    rgb[C[u]] += 1

    for w in u.adiacenti:
        if(w == v):
            #c'è almeno un arco rosso, verde e blu ?
            return (rgb['red'] > 0 and rgb['green'] > 0 and rgb['blue'] > 0)

        if(P[w] == -1):
            P[w] = v
            if DFS_color(G,C,P,rgb,u,v):
                return True

    rgb[C[u]] -=1
    
    return False
```

**Reply by Elia-Belli** [reply]
**Controesempio alla Correttezza**: nel seguente grafo se il DFS costruisce l'arborescenza seguendo i numeri degli archi in rosso, non verrà mai percorso il cammino con tutti i colori perché un nodo che ne fa parte è già stato visitato. 
<p align = center>
<img height = 250 src = https://github.com/user-attachments/assets/2c044f63-4141-4fe0-9606-76d210467600>
</p>
(Si potrebbe pensare di eliminare il nodo dal vettore dei padri prima di tornare dalla chiamata ricorsiva oltre ad effettuare il decremento del colore, in modo da poter rivisitare il nodo, ma andrebbero fatti vari accorgimenti e la complessità non sarebbe più quella del DFS)

**Comment by i0nut02** [comment]
```py
def visit(G, u, y, visited, combs):
    if visited[u]:
        return
    visited[u] = True

    if u == y:
        combs[y].append(G[y][0])
        return
    
    for v in G[u][1]:
        visit(G, v, y, visited, combs)
        # date le possibili visite con vari colori dati da visit di v,
        # aggiungiamo a quest'ultimi il colore del nodo attuale e vediamo
        # se c'e' gia, inoltre le possibili combinazioni sono 7 considerando
        # che il numero binario e' fatto di 3 cifre e b000 non esiste (cammino senza nodi)
        for comb in combs[v]: # O(7) = O(1)
            if not ((comb | G[u][0]) in combs[u]):
                combs[u].append(comb | G[u][0])
    return
    
def es(G, x, y): 
    # assumiamo che G sia del tipo List<Tuple<int, List<Int>>>
    # tipo [(1, [1, ...]), (2, [2, ...]), (4, [...]), ...]
    # e i colori saranno 1, 2, 3 che in binario saranno b001, b010, b100
    n = len(G)
    visited = [False for _ in range(n)]
    combs = [[] for _ in range(n)]

    visit(G, x, y, visited, combs)

    return 7 in combs[x] # vedi se tra i cammini di comb x c'e' uno con tutti i colori b111
```

Si potrebbe usare un set ma i prof non vogliono anche se qui non sarebbe un problemma l'inserimento in tempo lineare rispetto la lunghezza

**Comment by CiottoloMaggico** [comment]
```python
def dfs_mod(G, vis, vis_y, colors, v):
    cols = {"r": 0, "g": 1, "b":2}
    vis[v] = 1
    colors[cols[G.nodes[v]["color"]]] += 1

    for a in G.neighbors(v):
        if vis[a] == -1:
            dfs_mod(G, vis, vis_y, colors, a)
        
        if vis_y[a] == 1:
            vis_y[v] = 1

def es_4(G : nx.DiGraph, x, y):
    vis, vis_y, colors = [-1]*len(G), [-1]*len(G), [0,0,0]
    vis_y[y] = 1
    dfs_mod(G, vis, vis_y, colors, x)

    if vis_y[x] != 1 or (colors[0] == 0 or colors[1] == 0 or colors[2] == 0):
        return 0
    
    sorted_order = G.topological_sort()
    for n in sorted_order:
        if G.nodes[x]["color"] == G.nodes[y]["color"]:
            if vis_y[n] == 1 and G.nodes[n]["color"] != G.nodes[x]["color"]:
                return es_4(G, n, y)
        else:
            if vis_y[n] == 1 and G.nodes[n]["color"] != G.nodes[x]["color"] and G.nodes[n]["color"] != G.nodes[y]["color"]:
                return 1
            
    return 0
```

**Reply by i0nut02** [reply]
il costo di base non e' O(n + m) ma piu' che altro O(n(n + m)), considera che non e' detto che n sia visitato da x cioe ci potrebbe essere una cosa del genere x -> a -> b -> y e i_1 -> .... -> i_{n} quindi quando vai nel grafo delle i fai delle dfs e le fai all'incirca n volte, per questo il costo detto prima, inoltre e' anche sbagliato per cose come x (red) -> y (blue) e i (green) -> y (blue)
* dove ho messo "e" tra due grafi e' perche sono separati ma si trovano tutti in G

**Comment by niccolo-fato** [comment]
```py
def esercizio_3(G, u, v):  # G è il grafo colorato, u e v sono i vertici
    Vis = [0 for _ in range(len(G))]
    cr = cb = cv = 0

    if u == 'blue':
        cb += 1
    elif u == 'red':
        cr += 1
    else:
        cv += 1

    if u == v:
        return 'NO'

    return DFS(Vis, u, v, cr, cb, cv, False)

def DFS(Vis, u, target, cr, cb, cv, result):
    Vis[u] = 1

    for y in archi_uscenti(u):
        if y == target:
            if y == 'blue':
                cb += 1
            elif y == 'red':
                cr += 1
            else:
                cv += 1

            if cb >= 1 and cr >= 1 and cv >= 1:
                return 'YES'
            else:
                return 'NO'

        if Vis[y] == 0:
            if y == 'blue':
                result = result or DFS(Vis, y, target, cr, cb + 1, cv, result)
            elif y == 'red':
                result = result or DFS(Vis, y, target, cr + 1, cb, cv, result)
            else:
                result = result or DFS(Vis, y, target, cr, cb, cv + 1, result)

    return result

```

**Reply by i0nut02** [reply]
Questa soluzione e' sbagliata, basta vedere il caso che Elia Belli ha mostrato nel suo algoritmo

**Reply by niccolo-fato** [reply]
Ah ok, grazie mille

**Comment by Near047** [comment]
Idea, domani c'è l'appello quindi non scrivo codice. 
Si crea un Array [{000},....], il primo 0 per rosso, secondo per blu e terzo per verde. 
Se mi trovo al vertice `v` blu e prima di me ci stava solo `u`, un vertice rosso allora il mio A[i] sarà 110, se invece fosse stato blu anche  `u`, allora A[i] rimaneva 010, appena mi trovo ad un bivio dove `k` e `j` puntano ad `u`  e A[k] = 110, A[j] = 101 e `u` è rosso, allora scelgo quello che mi farà arrivare più velocemente 111, che propago.
Se A[vertice che vogliamo] è 111 allora SI. In tutto ciò userei un BFS che parte dal vertice dato. 
Spero abbia senso e che possa funzionare.

**Reply by i0nut02** [reply]
Non funziona perche' in realta' ti devi salvare sia 110 che 101, e tutte le combinazioni possibili della string "XXX" dove X $\in \{0, 1}$, sono 8, quindi tempo costante, ho scritto una soluzione con codice dove hai una array A = [["000"] for _ in range(|V|)], e dentro questo array ti salvi tutte le combinazioni che hai con il nodo i.

**Comment by cek753** [comment]
<img width="1353" height="892" alt="image" src="https://github.com/user-attachments/assets/6d5142b0-f107-4a85-bcb9-3a44aff02b03" />
